### ISSUE TITLE
Getting around Optional.get() for Simulator3.java

### ISSUE THREAD

**Issue by gordengorden** [nus-cs2030/2122-s2#93]
Hi guys, was wondering how you guys have been getting past not using Optional.get() to retrieve the Events to add into the PQ. So far some ideas I've thought of are just having the PQ contain Optional<Event>'s  but that doesn't seem very elegant. Interested to hear what some of you guys came up with!

**Comment by jingyulim1** [comment]
I used orElseThrow() to get the element, but only when I know that the element exists

**Comment by Yang990121** [comment]
Maybe can use orElse(___)

**Comment by sinteary** [comment]
Hi, just wondering what's the issue with using `Optional.get()`?

**Comment by TohDeKai** [comment]
> Hi, just wondering what's the issue with using `Optional.get()`?

i think the problem is that if you don't use it with a isPresent(), there's the possibility that you may get a null value and that's not ideal. 

However, I think that the isPresent() and get() combo is not a good practice as mentioned in the lecture, so just try to avoid it.

**Comment by sinteary** [comment]
Hi, I would like to clarify a few things about the suggestions given:
For @jingyulim1 's suggestion, since you mentioned that you only use if "when you know the element exists", I assume you still need to check `ifPresent()`, then access the element via `orElseThrow()`. But doesn't this mean that it ultimately works in a similar way to using `ifPresent()` and `get()`?
For @Yang990121 's suggestion, does this mean we have to use `orElse()` return a "dummy", and then check if it is a dummy? Doesn't this mean that we need to alter the base classes to suit the handling of Optionals?
Also wondering if anyone else has found other workarounds for this issue.
Thank you!

**Comment by jingyulim1** [comment]
Yep, I had another method to check if an Optional is present (since ifPresent() is banned, I checked if the Optional object == Optional.empty()). But since code design is graded for the project, I think one shouldn't try to come up with ways to bypass the checkers (dummies as well), but instead try to design around the issue.

**Comment by sinteary** [comment]
@jingyulim1 what you're suggesting seems more like a syntax issue rather than a design issue. Logically `isPresent()` and `object == Optional.empty()` seems to give the same boolean value. In fact I would argue that using Optional's API is better than having to manually check this yourself.  I don't see a significant difference in the way you are handling this compared with the combination of `isPresent()` and `get()` other than you are using different syntax. 
Btw did they really explicitly ban `isPresent()`?  I don't recall seeing this anywhere but please do correct me if I'm wrong.

**Comment by Yang990121** [comment]
Slide 4 of lecture 9
I believe previously ni the forum someone tried that in Lab5 and codecrunch didnt allow also

**Comment by sinteary** [comment]
@Yang990121 I see, thanks! In that case it seems inevitable that we should change to alternative syntax to overcome this problem. But it really begs the question whether banning such syntax truly solves the problem or not, or if the problem can be solved in the first place.
In this case, I think it's better to use orElse(*some default*) as you have suggested. Thank you so much!

**Comment by Yang990121** [comment]
If we didn't use our own PQ and use the Java's Priority queue, which is mutable, we could have used ifPresentOrElse to solve it, but we cant do that now cause ifPresentOrElse returns a void, so I also don't really know how to solve this issue properly. 
I think lab 6 LVL 4 u will face the same problem again

**Comment by sinteary** [comment]
For Lab 6 level 4, I used a combination of `flatMap()`, `map()` and `orElse()` to get the desired result. In this case it is defaulting to some predetermined value if it does not exist.

**Comment by Yang990121** [comment]
@sinteary did u use nested flatMap()  to get the Optional<student>? I cant seem to figure out how to use flatMap and Map()

**Comment by sinteary** [comment]
@Yang990121 Hi, yes I did. I think they gave a few examples in the lab task itself, perhaps you could refer to them!
